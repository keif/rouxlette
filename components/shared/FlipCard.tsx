import React, { useState } from 'react';
import { ViewStyle } from 'react-native';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  interpolate,
  runOnJS,
} from 'react-native-reanimated';
import { TapGestureHandler, PanGestureHandler, State } from 'react-native-gesture-handler';

interface FlipCardProps {
  front: React.ReactNode;
  back: React.ReactNode;
  flipped?: boolean;
  onFlipChange?: (flipped: boolean) => void;
  style?: ViewStyle;
  flipDurationMs?: number;
}

const FlipCard = ({
  front,
  back,
  flipped: controlledFlipped,
  onFlipChange,
  style,
  flipDurationMs = 300,
}: FlipCardProps) => {
  const [isFlipped, setIsFlipped] = useState(false);
  const flipValue = useSharedValue(0);
  
  const currentFlipped = controlledFlipped !== undefined ? controlledFlipped : isFlipped;
  
  const handleFlip = () => {
    const newFlippedState = !currentFlipped;
    
    if (controlledFlipped === undefined) {
      setIsFlipped(newFlippedState);
    }
    
    if (onFlipChange) {
      onFlipChange(newFlippedState);
    }
    
    flipValue.value = withTiming(newFlippedState ? 1 : 0, {
      duration: flipDurationMs,
    });
  };

  const onTapStateChange = (event: any) => {
    if (event.nativeEvent.state === State.END) {
      runOnJS(handleFlip)();
    }
  };

  const onPanStateChange = (event: any) => {
    if (event.nativeEvent.state === State.END) {
      const { translationX } = event.nativeEvent;
      // Trigger flip on horizontal swipe (threshold: 50 pixels)
      if (Math.abs(translationX) > 50) {
        runOnJS(handleFlip)();
      }
    }
  };

  const frontAnimatedStyle = useAnimatedStyle(() => {
    const rotateY = interpolate(flipValue.value, [0, 1], [0, 180]);
    const opacity = interpolate(flipValue.value, [0, 0.5, 1], [1, 0, 0]);
    
    return {
      transform: [{ perspective: 1000 }, { rotateY: `${rotateY}deg` }],
      opacity,
      backfaceVisibility: 'hidden',
      pointerEvents: flipValue.value > 0.5 ? 'none' : 'auto',
    };
  });

  const backAnimatedStyle = useAnimatedStyle(() => {
    const rotateY = interpolate(flipValue.value, [0, 1], [180, 0]);
    const opacity = interpolate(flipValue.value, [0, 0.5, 1], [0, 0, 1]);
    
    return {
      transform: [{ perspective: 1000 }, { rotateY: `${rotateY}deg` }],
      opacity,
      backfaceVisibility: 'hidden',
      pointerEvents: flipValue.value < 0.5 ? 'none' : 'auto',
    };
  });

  React.useEffect(() => {
    flipValue.value = withTiming(currentFlipped ? 1 : 0, {
      duration: flipDurationMs,
    });
  }, [currentFlipped, flipDurationMs]);

  return (
    <TapGestureHandler onHandlerStateChange={onTapStateChange}>
      <PanGestureHandler onHandlerStateChange={onPanStateChange}>
        <Animated.View
          style={[
            {
              position: 'relative',
            },
            style,
          ]}
          accessible={true}
          accessibilityRole="button"
          accessibilityHint="Flip card to see more details"
        >
          <Animated.View
            style={[
              {
                position: 'absolute',
                width: '100%',
                height: '100%',
              },
              frontAnimatedStyle,
            ]}
          >
            {front}
          </Animated.View>
          <Animated.View
            style={[
              {
                position: 'absolute',
                width: '100%',
                height: '100%',
              },
              backAnimatedStyle,
            ]}
          >
            {back}
          </Animated.View>
          {/* Invisible view to maintain height */}
          <Animated.View style={{ opacity: 0 }}>
            {front}
          </Animated.View>
        </Animated.View>
      </PanGestureHandler>
    </TapGestureHandler>
  );
};

export default FlipCard;