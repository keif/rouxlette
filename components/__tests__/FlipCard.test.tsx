import * as React from 'react';
import renderer from 'react-test-renderer';
import { Text, View } from 'react-native';

import FlipCard from '../shared/FlipCard';

// Mock react-native-reanimated
jest.mock('react-native-reanimated', () => {
  const Reanimated = require('react-native-reanimated/mock');
  Reanimated.default.call = () => {};
  return Reanimated;
});

// Mock react-native-gesture-handler
jest.mock('react-native-gesture-handler', () => {
  const mockReact = require('react');
  const original = jest.requireActual('react-native-gesture-handler');
  return {
    ...original,
    TapGestureHandler: ({ children }: any) => children,
    PanGestureHandler: ({ children }: any) => children,
    State: {
      END: 4,
    },
  };
});

const mockFront = React.createElement(View, { testID: 'front' }, 
  React.createElement(Text, null, 'Front Content')
);

const mockBack = React.createElement(View, { testID: 'back' },
  React.createElement(Text, null, 'Back Content')
);

it('renders FlipCard correctly', () => {
  const tree = renderer.create(
    React.createElement(FlipCard, { front: mockFront, back: mockBack })
  ).toJSON();

  expect(tree).toMatchSnapshot();
});