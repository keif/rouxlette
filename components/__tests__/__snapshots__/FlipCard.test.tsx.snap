// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FlipCard correctly 1`] = `
<View
  accessibilityHint="Flip card to see more details"
  accessibilityRole="button"
  accessible={true}
  style={
    [
      {
        "position": "relative",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "height": "100%",
          "position": "absolute",
          "width": "100%",
        },
        {
          "backfaceVisibility": "hidden",
          "opacity": undefined,
          "pointerEvents": "auto",
          "transform": [
            {
              "perspective": 1000,
            },
            {
              "rotateY": "undefineddeg",
            },
          ],
        },
      ]
    }
  >
    <View
      testID="front"
    >
      <Text>
        Front Content
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "height": "100%",
          "position": "absolute",
          "width": "100%",
        },
        {
          "backfaceVisibility": "hidden",
          "opacity": undefined,
          "pointerEvents": "none",
          "transform": [
            {
              "perspective": 1000,
            },
            {
              "rotateY": "undefineddeg",
            },
          ],
        },
      ]
    }
  >
    <View
      testID="back"
    >
      <Text>
        Back Content
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "opacity": 0,
      }
    }
  >
    <View
      testID="front"
    >
      <Text>
        Front Content
      </Text>
    </View>
  </View>
</View>
`;
